server/datastore/mysql/testing_utils.go

```diff
@@ -0,0 +1,129 @@
+	testUsername = "root"
+	testPassword = "toor"
+	testAddress  = "localhost:3307"
+)
+
+func connectMySQL(t *testing.T, testName string) *Datastore {
+	config := config.MysqlConfig{
+		Username: testUsername,
+		Password: testPassword,
+		Database: testName,
+		Address:  testAddress,
+	}
+
+	// Create datastore client
+	ds, err := New(config, clock.NewMockClock(), Logger(log.NewNopLogger()), LimitAttempts(1))
+	require.Nil(t, err)
+	return ds
+}
+
+// initializeSchema initializes a database schema using the normal Fleet
+// migrations, then outputs the schema with mysqldump within the MySQL Docker
+// container.
+func initializeSchema(t *testing.T) {
+	// Create the database (must use raw MySQL client to do this)
+	db, err := sql.Open(
+		"mysql",
+		fmt.Sprintf("%s:%s@tcp(%s)/?multiStatements=true", testUsername, testPassword, testAddress),
+	)
+	require.NoError(t, err)
+	defer db.Close()
+	_, err = db.Exec("DROP DATABASE IF EXISTS schemadb; CREATE DATABASE schemadb;")
+	require.NoError(t, err)
+
+	// Create a datastore client in order to run migrations as usual
+	config := config.MysqlConfig{
+		Username: testUsername,
+		Password: testPassword,
+		Address:  testAddress,
+		Database: schemaDbName,
+	}
+	ds, err := New(config, clock.NewMockClock(), Logger(log.NewNopLogger()), LimitAttempts(1))
+	require.Nil(t, err)
+	defer ds.Close()
+	require.Nil(t, ds.MigrateTables())
+
+	// Dump schema to dumpfile
+	if out, err := exec.Command(
+		"docker-compose", "exec", "-T", "mysql_test",
+		// Command run inside container
+		"mysqldump",
+		"-u"+testUsername, "-p"+testPassword,
+		"schemadb",
+		"--compact", "--skip-comments",
+		"--result-file="+dumpfile,
+	).CombinedOutput(); err != nil {
+		t.Error(err)
+		t.Error(string(out))
+		t.FailNow()
+	}
+}
+
+// initializeDatabase loads the dumped schema into a newly created database in
+// MySQL. This is much faster than running the full set of migrations on each
+// test.
+func initializeDatabase(t *testing.T, dbName string) {
+	// Load schema from dumpfile
+	if out, err := exec.Command(
+		"docker-compose", "exec", "-T", "mysql_test",
+		// Command run inside container
+		"mysql",
+		"-u"+testUsername, "-p"+testPassword,
+		"-e",
+		fmt.Sprintf(
+			"DROP DATABASE IF EXISTS %s; CREATE DATABASE %s; USE %s; SET FOREIGN_KEY_CHECKS=0; SOURCE %s;",
+			dbName, dbName, dbName, dumpfile,
+		),
+	).CombinedOutput(); err != nil {
+		t.Error(err)
+		t.Error(string(out))
+		t.FailNow()
+	}
+
+}
+
+func runTest(t *testing.T, testFunc func(*testing.T, fleet.Datastore)) {
+	t.Run(test.FunctionName(testFunc), func(t *testing.T) {
+		t.Parallel()
+
+		// Create a new database and load the schema for each test
+		initializeDatabase(t, test.FunctionName(testFunc))
+
+		ds := connectMySQL(t, test.FunctionName(testFunc))
+		defer ds.Close()
+
+		testFunc(t, ds)
+	})
+}
+
+func RunTestsAgainstMySQL(t *testing.T, tests []func(*testing.T, fleet.Datastore)) {
+	if _, ok := os.LookupEnv("MYSQL_TEST"); !ok {
```

Comment:
```suggestion
func RunTestsAgainstMySQL(t *testing.T, tests []func(*testing.T, fleet.Datastore)) {
	t.Parallel()
	if _, ok := os.LookupEnv("MYSQL_TEST"); !ok {
```

Reply:
We use `t.Parallel()` later on right after we do `t.Run`: http://github.com/fleetdm/fleet/blob/3b5edb92a82145eb0dad3855655e3f7594299163/server/datastore/mysql/testing_utils.go#L106-L106

Reply:
I may be misinterpreting, but I think if we don't call `t.Parallel()` here or in https://github.com/fleetdm/fleet/pull/1322/files#diff-7dbb80f5da11b554e78c62c9b0dcdfa41c11952b3043f83501e64763d48245e6R10, the subtests will be allowed to run in parallel with each other, but the top level test will run sequentially.

Now that we are calling `RunTestsAgainstMySQL` from multiple packages this may actually be the desired behavior, as the `initializeSchema`/`initializeDatabase` pattern is not designed to handle concurrent executions.